pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,249 @@
+// the batch used in the postage stamp attached to the chunk.
+// Depending on the value of the batch (reserve depth of the batch), the localstore
+// either pins the chunk (thereby protecting it from garbage collection) or not.
+// the chunk stays pinned until it is 'unreserved' based on changes in relative priority of the batch it belongs to
+//
+// the atomic db operation is unreserving a batch down to a depth
+// the intended semantics of unreserve is to unpin the chunks
+// in the relevant POs, belonging to the batch and (unless they are otherwise pinned)
+// allow  them  to be gargage collected.
+//
+// the rules of the reserve
+// - if batch a is unreserved and val(b) <  val(a) then b is unreserved on any po
+// - if a batch is unreserved on po  p, then  it is unreserved also on any p'<p
+// - batch size based on fully filled the reserve should not  exceed Capacity
+// - batch reserve is maximally utilised, i.e, cannot be extended and have 1-3 remain true
+package batchstore
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// ErrBatchNotFound is returned when the postage batch is not found or expired
+var ErrBatchNotFound = errors.New("postage batch not found or expired")
+
+// DefaultDepth is the initial depth for the reserve
+const DefaultDepth = 5
+
+// Capacity = number of chunks in reserve. `2^23` was chosen to remain
+// relatively near the current 5M chunks ~25GB
+var Capacity = exp2(23)
+
+var big1 = big.NewInt(1)
+
+// reserveState records the state and is persisted in the state store
+type reserveState struct {
+	Depth    uint8    `json:"depth"`    // Radius of responsibility
+	Capacity int64    `json:"capacity"` // size of the reserve = number of chunks
+	Outer    *big.Int `json:"outer"`    // lower value limit for outer layer = the further half of chunks
+	Inner    *big.Int `json:"inner"`    // lower value limit for inner layer = the closer half of chunks
+}
+
+// unreserve is called when the batchstore decides not to reserve a batch on a PO
+// i.e.  chunk of  the batch in bins [0 upto PO] (closed  interval) are unreserved
+func (s *store) unreserve(b *postage.Batch, radius uint8) error {
+	return s.unreserveFunc(b.ID, radius)
+}
+
+// evictExpired is called when PutChainState is called (and there is 'settlement')
+func (s *store) evictExpired() error {
+	var toDelete [][]byte
+	// set until to total or inner whichever is greater
+	until := new(big.Int)
+	if s.rs.Inner.Cmp(big.NewInt(0)) > 0 && s.cs.Total.Cmp(s.rs.Inner) >= 0 {
+		until.Add(s.cs.Total, big1)
+	} else {
+		until.Set(s.rs.Inner)
+	}
+	var multiplier int64
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		b, err := s.Get(valueKeyToID(key))
+		if err != nil {
+			return true, err
+		}
+		// terminate iteration if until is passed
+		if b.Value.Cmp(until) >= 0 {
+			return true, nil
+		}
+		//
```

Comment:
not sure what these if statements signify. it seems documentation started but has not completed